File: packages/javascript/core/context.js
/**
 * Context: The Loving Vessel
 *
 * With agape compassion, the Context holds data tenderly, immutable by default for safety, mutable for flexibility.
 * Optimized for JavaScript's dynamism—embracing object-like interface with ecosystem integrations.
 */

class Context {
  /**
   * Immutable context with selfless love—holds data without judgment, returns fresh copies for changes.
   * @param {Object} data - Initial data object
   */
  constructor(data = {}) {
    this._data = this._deepFreeze({ ...data });
  }

  /**
   * Deep freeze an object to ensure immutability at all levels
   * @param {Object} obj - The object to deep freeze
   * @returns {Object} The deep frozen object
   */
  _deepFreeze(obj) {
    if (obj === null || typeof obj !== 'object') return obj;
    
    // Freeze the object
    Object.freeze(obj);
    
    // Recursively freeze all properties
    Object.keys(obj).forEach(key => {
      if (typeof obj[key] === 'object' && obj[key] !== null && !Object.isFrozen(obj[key])) {
        this._deepFreeze(obj[key]);
      }
    });
    
    return obj;
  }

  /**
   * Create an empty context
   * @returns {Context} An empty context
   */
  static empty() {
    return new Context({});
  }

  /**
   * Create a context from data
   * @param {Object} data - The data to create context from
   * @returns {Context} A new context with the data
   */
  static from(data) {
    return new Context(data);
  }

  /**
   * With gentle care, return the value or undefined, forgiving absence.
   * Returns a deep copy of complex objects to maintain immutability.
   * @param {string} key - The key to retrieve
   * @returns {*} The value or undefined
   */
  get(key) {
    const value = this._data[key];
    if (value === undefined) return undefined;
    
    // Return deep copy for objects and arrays to maintain immutability
    if (typeof value === 'object' && value !== null) {
      return JSON.parse(JSON.stringify(value));
    }
    
    return value;
  }

  /**
   * With selfless safety, return a fresh context with the addition.
   * @param {string} key - The key to insert
   * @param {*} value - The value to insert
   * @returns {Context} A new Context with the addition
   */
  insert(key, value) {
    const newData = { ...this._data, [key]: value };
    return new Context(newData);
  }

  /**
   * For those needing change, provide a mutable sibling.
   * @returns {MutableContext} A mutable version of this context
   */
  withMutation() {
    return new MutableContext({ ...this._data });
  }

  /**
   * Lovingly combine contexts, favoring the other with compassion.
   * @param {Context} other - The other context to merge
   * @returns {Context} A new Context with merged data
   */
  merge(other) {
    const newData = { ...this._data, ...other._data };
    return new Context(newData);
  }

  /**
   * Express as plain object for ecosystem integration.
   * @returns {Object} A copy of the internal data
   */
  toObject() {
    return JSON.parse(JSON.stringify(this._data));
  }

  /**
   * Check if a key exists in the context.
   * @param {string} key - The key to check
   * @returns {boolean} True if the key exists
   */
  has(key) {
    return key in this._data;
  }

  /**
   * Get all keys in the context.
   * @returns {string[]} Array of keys
   */
  keys() {
    return Object.keys(this._data);
  }

  toString() {
    return `Context(${JSON.stringify(this._data)})`;
  }
}

class MutableContext {
  /**
   * Mutable context for performance-critical sections—use with care, but forgiven.
   * @param {Object} data - Initial data object
   */
  constructor(data = {}) {
    this._data = { ...data };
  }

  /**
   * Get a value from the context.
   * @param {string} key - The key to retrieve
   * @returns {*} The value or undefined
   */
  get(key) {
    return this._data[key];
  }

  /**
   * Change in place with gentle permission.
   * @param {string} key - The key to set
   * @param {*} value - The value to set
   */
  set(key, value) {
    this._data[key] = value;
  }

  /**
   * Return to safety with a fresh immutable copy.
   * @returns {Context} An immutable Context
   */
  toImmutable() {
    return new Context(this._data);
  }

  /**
   * Check if a key exists.
   * @param {string} key - The key to check
   * @returns {boolean} True if the key exists
   */
  has(key) {
    return key in this._data;
  }

  /**
   * Get all keys.
   * @returns {string[]} Array of keys
   */
  keys() {
    return Object.keys(this._data);
  }

  toString() {
    return `MutableContext(${JSON.stringify(this._data)})`;
  }
}

module.exports = { Context, MutableContext };

File: packages/rust/src/core/context.rs
/*!
State: The Data Container

The State holds data carefully, immutable by default for safety, mutable for flexibility.
Optimized for Rust's ownership model—embracing HashMap with serde integration.
*/

use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

/// Immutable state with selfless love—holds data without judgment, returns fresh copies for changes.
/// Generic type parameter T represents the current data shape, defaulting to serde_json::Value for flexibility.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct State<T = Value> {
    data: HashMap<String, Value>,
    _phantom: std::marker::PhantomData<T>,
}

impl<T> State<T> {
    /// Create a new state with optional initial data
    pub fn new(data: HashMap<String, Value>) -> Self {
        Self {
            data,
            _phantom: std::marker::PhantomData,
        }
    }

    /// Create an empty state
    pub fn empty() -> Self {
        Self {
            data: HashMap::new(),
            _phantom: std::marker::PhantomData,
        }
    }

    /// With gentle care, return the value or None, forgiving absence.
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.data.get(key)
    }

    /// With selfless safety, return a fresh state with the addition (preserves type).
    pub fn insert(self, key: String, value: Value) -> State<T> {
        let mut new_data = self.data;
        new_data.insert(key, value);
        State {
            data: new_data,
            _phantom: std::marker::PhantomData,
        }
    }

    /// Type evolution: Transform to a new type while preserving data.
    pub fn insert_as<U>(self, key: String, value: Value) -> State<U> {
        let mut new_data = self.data;
        new_data.insert(key, value);
        State {
            data: new_data,
            _phantom: std::marker::PhantomData,
        }
    }

    /// For those needing change, provide a mutable sibling.
    pub fn with_mutation(&self) -> MutableState {
        MutableState {
            data: self.data.clone(),
        }
    }

    /// Lovingly combine states, favoring the other with compassion.
    pub fn merge(mut self, other: &State<T>) -> State<T> {
        for (key, value) in &other.data {
            self.data.insert(key.clone(), value.clone());
        }
        self
    }

    /// Express as HashMap for ecosystem integration.
    pub fn to_hashmap(&self) -> HashMap<String, Value> {
        self.data.clone()
    }

    /// Get a reference to the internal data
    pub fn data(&self) -> &HashMap<String, Value> {
        &self.data
    }
}

impl State<Value> {
    /// Create state from HashMap (for backward compatibility)
    pub fn from_hashmap(data: HashMap<String, Value>) -> Self {
        Self::new(data)
    }
}

impl<T> Default for State<T> {
    fn default() -> Self {
        Self::empty()
    }
}

/// Mutable state for performance-critical sections—use with care, but forgiven.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MutableState {
    data: HashMap<String, Value>,
}

impl MutableState {
    /// Create a new mutable state
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
        }
    }

    /// With gentle care, return the value or None
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.data.get(key)
    }

    /// Change in place with gentle permission.
    pub fn set(&mut self, key: String, value: Value) {
        self.data.insert(key, value);
    }

    /// Return to safety with a fresh immutable copy.
    pub fn to_immutable<T>(self) -> State<T> {
        State {
            data: self.data,
            _phantom: std::marker::PhantomData,
        }
    }

    /// Get a reference to the internal data
    pub fn data(&self) -> &HashMap<String, Value> {
        &self.data
    }
}

impl Default for MutableState {
    fn default() -> Self {
        Self::new()
    }
}